ffi/src/handle.rs

```diff
@@ -58,15 +58,16 @@ macro_rules! def_ffi_fn {
             },
         }
     };
-    ( $vis:vis Clone: $( $other:ty ),+ $(,)? ) => {
+    ( $($prefix:literal)? Clone: $( $other:ty ),+ $(,)? ) => {
         /// FFI function equivalent of [`Clone::clone`]
         ///
         /// # Safety
         ///
         /// All of the given pointers must be valid and the given handle id must match the expected
         /// pointer type
         #[no_mangle]
-        $vis unsafe extern "C" fn __clone(
+        $(#[export_name = concat!($prefix, "__clone")])?
+        unsafe extern "C" fn __clone(
```

Comment:
current behavior is such that one `__clone` is generated per crate because of orphan rule and to avoid the risk of collision of `Handle::ID`s across crates. Since there are multiple versions of `__clone` they are disambiguated by adding a prefix. Currently handle id is unique per crate

It would be possible to generate only one `__clone` across all crates but that approach has the risk of two structs from different crates sharing the same `Handle::ID` by accident resulting in UB because of invalid void pointer cast. In that case we would have to make sure that id is globally unique across all crates

Reply:
1. I'd like this in the text of the program, preferably as a doc-string. 
2. I'd like a couple of unit tests that show us how to break this invariant. 
3. Hard-coding the export name is **much** better because it decouples the module path from the actual implementation. With that said, I'd rather we had a more stable system.

Reply:
> 1. I'd like this in the text of the program, preferably as a doc-string. 

I agree

> 2. I'd like a couple of unit tests that show us how to break this invariant.

Maybe I can make a test for this. This would be testing that linker refuses to link two crates that have exported a function with the same name.

> 3. Hard-coding the export name is **much** better because it decouples the module path from the actual implementation. With that said, I'd rather we had a more stable system. 

What would a "more stable" system be?

Reply:
Sorry, I was confused. The doc should specifically contrast the brittle approach (one global `__clone`) with the current implementation per-crate `__clone`